klog.go

```diff
@@ -862,7 +862,7 @@ func (rb *redirectBuffer) Write(bytes []byte) (n int, err error) {
 // Use as:
 //   ...
 //   klog.SetLogger(zapr.NewLogger(zapLog))
-func SetLogger(logr logr.Logger) {
+func SetLogger(logr *logr.Logger) {
```

Comment:
I'm not convinced you need to take a pointer here.  Copying the logr.Logger struct should be cheap.  Then you can avoid this API break.

Reply:
Code like https://github.com/kubernetes/klog/blob/69472594defe33aee4442f9266aadfddc89fe4b0/klog.go#L782-L785 must be able to check whether a logger has been set. That could also be done by checking for non-nil logging.logr.LogSink, though.

But API break is an interesting topic - isn't it also an API break if code using a logr.Logger interface (i.e. logr < 1.0.0-rc1) can no longer call SetLogger?

Reply:
Or do we argue that the semantic versioning of klog does not apply to those parts of it which use types from dependencies which themselves had no API stability guarantee yet?

Reply:
Ahh, well, you could internally store a pointer but accept a value.

Reply:
That's still an API change because previously valid values (interface) become invalid (now need the struct). The breakage is rather indirect because whatever code provided the logger implementation will also no longer compile (we cannot have logr v0.4 and v1.0.0 at the same time), but it's still breakage caused by updating klog.

Having said that, I prefer passing the logger by value. My  reasoning is that code which just passes logger instances around (formerly interfaces, now structs) continues to work unchanged.

Reply:
switching the signature back

Reply:
why not change logr.Logger to logr.LogSink ?